colcon_bundle/verb/bundlefile.py

```diff
@@ -0,0 +1,107 @@
+import json
+import os
+import tarfile
+import tempfile
+
+from .utilities import filechecksum
+
+
+class BundleFile:
+    """Provides an interface to application bundle files."""
+
+    def __init__(self, *, name=None, mode='w'):
+        """
+        Open a bundle archive for writing.
+
+        mode:
+        'w' open for writing
+
+        :param name: path to the bundle archive
+        :param mode: mode to open file as
+        """
+        self.tarfile = tarfile.open(name, mode)
+        self.overlays = []
+        self.metadata = []
+        self.mode = mode
+
+    def add_metadata(self, path):
+        """
+        Add the json file at path to the metadata archive.
+
+        :param path: path to json file
+        """
+        self.metadata.append(path)
+
+    def add_overlay_archive(self, path):
+        """
+        Add the archive at path to the bundle as an overlay.
+
+        This does not write immediately, once the object is closed all
+        writes will occur.
+        """
+        self.overlays.append(path)
+
+    def close(self):
+        """Close the archive."""
+        max_data_size = 4 * 1024 * 1024
+        if 'w' in self.mode:
+            tempdir = tempfile.mkdtemp()
+            version_path = os.path.join(tempdir, 'version')
+
+            with open(version_path, 'w') as v:
+                v.write('2')
+            self.tarfile.add(version_path, arcname='version')
+
+            offset = max_data_size
+            overlay_metadata = []
+            for overlay in self.overlays:
+                name = os.path.basename(overlay)
+                checksum = filechecksum(overlay)
+                info = self.tarfile.gettarinfo(overlay, arcname=name)
+                header_size = len(info.tobuf())
+                file_size = os.stat(overlay).st_size
+                total = header_size + file_size
+                overlay_metadata.append({
+                    'name': name,
+                    'sha256': checksum,
+                    'offset': offset + header_size,
```

Comment:
Would the clients that download the bundle need to read header for the overlay?

Reply:
No, if they want to read headers they can read through the tar file. The offset is directly to the file contents.